amr-wind/physics/ConvectingTaylorVortex.cpp

```diff
@@ -0,0 +1,164 @@
+#include "amr-wind/physics/ConvectingTaylorVortex.H"
+#include "amr-wind/CFDSim.H"
+#include "AMReX_iMultiFab.H"
+#include "AMReX_MultiFabUtil.H"
+#include "AMReX_ParmParse.H"
+#include "amr-wind/utilities/trig_ops.H"
+
+namespace amr_wind {
+namespace ctv {
+
+namespace {
+amrex::Real u_exact(
+    const amrex::Real u0,
+    const amrex::Real v0,
+    const amrex::Real omega,
+    const amrex::Real x,
+    const amrex::Real y,
+    const amrex::Real t)
+{
+    return u0 - std::cos(utils::pi() * (x - u0 * t)) *
+                    std::sin(utils::pi() * (y - v0 * t)) *
+                    std::exp(-2.0 * omega * t);
+}
+
+amrex::Real v_exact(
+    const amrex::Real u0,
+    const amrex::Real v0,
+    const amrex::Real omega,
+    const amrex::Real x,
+    const amrex::Real y,
+    const amrex::Real t)
+{
+    return v0 + std::sin(utils::pi() * (x - u0 * t)) *
+                    std::cos(utils::pi() * (y - v0 * t)) *
+                    std::exp(-2.0 * omega * t);
+}
+
+} // namespace
+
+ConvectingTaylorVortex::ConvectingTaylorVortex(const CFDSim& sim)
+    : m_time(sim.time())
+    , m_repo(sim.repo())
+    , m_mesh(sim.mesh())
+    , m_velocity(sim.repo().get_field("velocity"))
+    , m_density(sim.repo().get_field("density"))
+{
+    amrex::ParmParse pp("CTV");
+    pp.query("density", m_rho);
+    pp.query("u0", m_u0);
+    pp.query("v0", m_v0);
+    {
+        amrex::Real nu;
+        amrex::ParmParse pp("transport");
+        pp.query("viscosity", nu);
+        m_omega = utils::pi() * nu;
+    }
+    if (amrex::ParallelDescriptor::IOProcessor()) {
+        std::ofstream f;
+        f.open(m_output_fname.c_str());
+        f << std::setw(m_w) << "time" << std::setw(m_w) << "L2_u"
+          << std::setw(m_w) << "L2_v" << std::endl;
+        f.close();
+    }
+}
+
+/** Initialize the velocity and density fields at the beginning of the
+ *  simulation.
+ */
+void ConvectingTaylorVortex::initialize_fields(
+    int level, const amrex::Geometry& geom)
+{
+    using namespace utils;
+
+    auto& velocity = m_velocity(level);
+    auto& density = m_density(level);
+
+    density.setVal(m_rho);
+
+    for (amrex::MFIter mfi(velocity); mfi.isValid(); ++mfi) {
+        const auto& vbx = mfi.validbox();
+
+        const auto& dx = geom.CellSizeArray();
+        const auto& problo = geom.ProbLoArray();
+        auto vel = velocity.array(mfi);
+
+        amrex::ParallelFor(
+            vbx, [=] AMREX_GPU_DEVICE(int i, int j, int k) noexcept {
+                const amrex::Real x = problo[0] + (i + 0.5) * dx[0];
+                const amrex::Real y = problo[1] + (j + 0.5) * dx[1];
+                vel(i, j, k, 0) = u_exact(m_u0, m_v0, m_omega, x, y, 0.0);
+                vel(i, j, k, 1) = v_exact(m_u0, m_v0, m_omega, x, y, 0.0);
+                vel(i, j, k, 2) = 0.0;
+            });
+    }
+}
+
+amrex::Real ConvectingTaylorVortex::compute_error(
```

Comment:
Seems like this can be a standalone function in this file, and reduce exposure of FuncDef into other places the header is included? Not a big deal as we don't include this header anywhere in amr-wind sources, only unit tests.